core/sile.lua

```diff
@@ -240,10 +241,251 @@ function SILE.resolveFile(filename, pathprefix)
+    line = line,
+    column = column,
+    options = options or {}
+  })
+end
+--- Push a command frame on to the stack to record the execution trace for debugging.
+--- Command arguments are inferred from AST content, any item may be overridden.
+--- Must be popped with `pop(returnOfPush)`.
+function SILE.currentCommandStack:pushContent(content, tag, line, column, options, file)
+  return self:_pushFrame({
+    tag = tag or (type(content) == "table" and content.tag) or "???",
+    file = file or (type(content) == "table" and content.file) or SILE.currentlyProcessingFile,
+    line = line or (type(content) == "table" and content.line),
+    column = column or (type(content) == "table" and content.col),
+    options = options or (type(content) == "table" and content.attr) or {}
+  })
+end
+--- Push a text that is going to get typeset on to the stack to record the execution trace for debugging.
+--- Must be popped with `pop(returnOfPush)`.
+function SILE.currentCommandStack:pushText(text, line, column, file)
+  return self:_pushFrame({
+    text = text,
+    file = file,
+    line = line,
+    column = column
+  })
+end
+--- Internal: Push complete frame
+function SILE.currentCommandStack:_pushFrame(frame)
+  frame.toString = self._frameToString
+  frame.typesetter = SILE.typesetter
+  --frame.pushTrace = debug.traceback(nil, 3)
+
+  -- Push the frame
+  if SU.debugging("commandStack") then
+    print(string.rep(" ", #self) .. "PUSH(" .. frame:toString(false, true) .. ")")
+  end
+  self[#self + 1] = frame
+  self.lastCommand = nil
+  self._lastPushId = self._lastPushId + 1
+  frame.pushId = self._lastPushId
+  return self._lastPushId
+end
+--- Pop previously pushed command from the stack.
+--- Return value of `push` function must be provided as argument to check for balanced usage.
+function SILE.currentCommandStack:pop(pushId)
+  if type(pushId) ~= "number" then
+    SU.error("SILE.currentCommandStack:pop's argument must be the result value of the corresponding push", true)
+  end
+
+  -- First verify that push/pop is balanced
+  local popped = self[#self]
+  if popped.pushId ~= pushId then
+    local message = "Unbalanced content push/pop"
+    local debug = SILE.detailedErrors or SU.debugging("commandStack")
+    if debug then
+      message = message .. ". Expected " .. popped.pushId .. " - (" .. popped:toString() .. "), got " .. pushId
+    end
+    SU.warn(message, debug)
+  else
+    -- Correctly balanced: pop the frame
+    self.lastCommand = popped
+    self[#self] = nil
+    if SU.debugging("commandStack") then
+      print(string.rep(" ", #self) .. "POP(" .. popped:toString(false, true) .. ")")
+    end
+  end
+end
+--- Returns a stack of commands, which can be used to show information about location in document.
+--- Most relevant object is in the last index.
+--- Second return: recently popped value which could provide extra location information
+function SILE.currentCommandStack:locationStack(currentTypesetterOnly)
```

Comment:
Function names like this should probably be verbs, maybe `:dumpStack()` or something like that.

Reply:
No, disagree. This is returning a value, which is fine. A later function actually dumps it.

Reply:
Okay right dump was the wrong term, but still what does the function _do_? See my other comment about the naming of this object, but along with that maybe this should be `getLocations()` ... or especially with the argument it should be called `filterSomething()`.

Reply:
Names of these things have evolved over the time, so they don't make perfect sense right now. The names have grown on me, so I didn't realize it. I'll think of a better names once the name for this whole structure is decided.